### PR TITLE
Product Compare / M2 Support

### LINKED ISSUE #1410: Product Compare / M2 Support
**Description**:
Compare Products generates a detailed, side-by-side comparison of two or more products.

**DESIGN**
mobile - zpl.io/VQjEX4P
desktop - zpl.io/VDEnEDr & zpl.io/VKRlR0q

**TODO**
- Create a compare page
- Add the compare icon in the header 
-- The link to the compare page should be added to the icon
- Style the page as per the design on mobile and desktop 
- Clear compare, price, and compare labels are translatable 
- Products cards and attributes are scrollable (check tdkz behavior on both mobile and desktop)
- Products can be removed from compare on click on X button
- Products can be added to cart on click on add to cart button 
- Product name and product image are linked with the respected products page
- On the clear compare button, remove all items from compare
- All customer groups can add items to compare 
- On mobile, attribute label is sticky and always visible  

**Expected result**

- Users can add products for comparison through PDP / PLP
- Comparison page created
- Any number of products can be added, but more than 4 introduces horizontal scroll bar as in TDKZ: https://www.technodom.kz/compare
- 2 products are visible on mobile, the third is partially visible to indicate scrolling ability
- Setting in Admin → Stores → Attributes → Product → {attribute} → Storefront properties → Comparable on Storefront is respected
- Able to compare configurable products, default variant is selected
- Header extended with icon leading to compare page. 
- Responsive

### PR DESCRIPTION
**Pull request by zans-laksa**
Fixes #1410 

Related PRs:
scandipwa/compare-graphql/pull/3
scandipwa/performance/pull/12

### CODE REVIEW

File: src/app/component/ProductCard/ProductCard.component.js [hunk 1]
```diff
@@ -283,16 +335,18 @@ export class ProductCard extends PureComponent {
         return (
             this.renderCardLinkWrapper((
                 <>
-                    <figure block="ProductCard" elem="Figure">
-                        { this.renderPicture() }
-                    </figure>
-                    <div block="ProductCard" elem="Content">
+                    <div block="ProductCard" elem="Figure-Review">
```
Comment (alfredsgenkins) [inline src/app/component/ProductCard/ProductCard.component.js:338]: This does not seem right, consider not using `_` and `-` in blocks, mods, elems.

File: src/app/component/ProductCompare/ProductCompare.component.js [hunk 2]
```diff
@@ -0,0 +1,152 @@
...
+                  elem="AttributeTable"
+                >
+                    { this.renderAttributes() }
+                </div>
+            </div>
+        );
+    }
+
+    renderEmpty() {
+        return <div block="ProductCompare" elem="Empty">{ __('You have nothing to compare') }</div>;
```
Comment (alfredsgenkins) [inline src/app/component/ProductCompare/ProductCompare.component.js:130]: Consider transforming into multiline expression instead.

File: src/app/component/ProductCompare/ProductCompare.component.js [hunk 3]
```diff
@@ -0,0 +1,152 @@
...
+
+        if (isLoading && !hasProducts) {
+            return null;
+        }
+
+        if (!hasProducts) {
+            return this.renderEmpty();
+        }
+
+        return this.renderProducts();
```
Comment (alfredsgenkins) [inline src/app/component/ProductCompare/ProductCompare.component.js:148]: You indended it to show product when loading, if not empty?

File: src/app/component/ProductCompare/ProductCompare.container.js [hunk 4]
```diff
@@ -0,0 +1,107 @@
...
+        isLoading: false,
+        products: []
+    };
+
+    containerFunctions = {
+        getAttributes: this.getAttributes.bind(this),
+        clearCompareList: this.clearCompareList.bind(this)
+    };
+
+    componentWillMount() {
```
Comment (alfredsgenkins) [inline src/app/component/ProductCompare/ProductCompare.container.js:61]: This is unsafe method, no? https://reactjs.org/docs/react-component.html#unsafe_componentwillmount

File: src/app/component/ProductCompare/ProductCompare.style.scss [hunk 5]
```diff
@@ -0,0 +1,147 @@
...
+:root {
+    --product-compare-area-width: var(--content-wrapper-width);
+    --product-compare-sidebar-width: 200px;
+    --product-compare-item-count: 3.2;
+
+    @include mobile {
+        --product-compare-area-width: 100vw;
+        --product-compare-sidebar-width: 0px;
+        --product-compare-item-count: 2.1;
+    }
```
Comment (alfredsgenkins) [inline src/app/component/ProductCompare/ProductCompare.style.scss:21]: Consider declaring mobile overrides after rest of the variables.

File: src/app/component/ProductCompareButton/ProductCompareButton.style.scss [hunk 6]
```diff
@@ -0,0 +1,65 @@
...
+        .ProductCompareButton-Button {
+            border: none;
+            background-color: var(--product-compare-button-grey-background);
+        }
+
+        .ProductCompareButton-Icon,
+        .ProductCompareButton-Icon::before,
+        .ProductCompareButton-Icon::after {
+            background-color: var(--product-compare-button-grey-color);
+        }
```
Comment (alfredsgenkins) [inline src/app/component/ProductCompareButton/ProductCompareButton.style.scss:63]: Please rewrite this, so:

```scss
.ProductCompareButton {
    &-Button {
        border: none;
        background-color: var(--product-compare-button-grey-background);
    }

    &-Icon {
        &,
        &::after,
        &::before {
            background-color: var(--product-compare-button-grey-color);
        }
    }
}
```

File: src/app/component/Slider/Slider.style.scss [hunk 7]
```diff
@@ -47,6 +56,15 @@
 
         &_isVertical {
             flex-direction: column;
+            width: 100%;
+            max-height: var(--slide-height);
```
Comment (alfredsgenkins) [inline src/app/component/Slider/Slider.style.scss:60]: How is this related to compare? :D

File: src/app/component/InstallPrompt/InstallPrompt.component.js [hunk 8]
```diff
@@ -51,14 +51,14 @@ export class InstallPrompt extends PureComponent {
     }
 
     renderPrompt() {
-        const { device } = this.props;
+        const { device, containerFunctions } = this.props;
```
Comment (alfredsgenkins) [inline src/app/component/InstallPrompt/InstallPrompt.component.js:54]: How is this related to compare?

File: src/app/store/LinkedProducts/LinkedProducts.dispatcher.js [hunk 9]
```diff
@@ -74,7 +74,7 @@ export class LinkedProductsDispatcher extends QueryDispatcher {
     prepareRequest(product_links) {
         const relatedSKUs = product_links.reduce((links, link) => {
             const { linked_product_sku } = link;
-            return [...links, `"${ linked_product_sku.replace(/ /g, '%20') }"`];
+            return [...links, `${ linked_product_sku.replace(/ /g, '%20') }`];
```
Comment (alfredsgenkins) [inline src/app/store/LinkedProducts/LinkedProducts.dispatcher.js:77]: How is that related to compare? :D

File: src/app/component/SliderWidget/SliderWidget.style.scss [hunk 10]
```diff
@@ -10,6 +10,20 @@
  */
 
 .SliderWidget {
+    &-Figure {
```
Comment (alfredsgenkins) [inline src/app/component/SliderWidget/SliderWidget.style.scss:13]: How is that related to compare?

File: src/app/component/SliderWidget/SliderWidget.container.js [hunk 11]
```diff
@@ -71,6 +72,12 @@ export class SliderWidgetContainer extends DataContainer {
     }
 
     render() {
+        const { isLoaded } = this.state;
+
+        if (!isLoaded) {
+            return null;
```
Comment (alfredsgenkins) [inline src/app/component/SliderWidget/SliderWidget.container.js:78]: Render `null` might cause jumps! How is that related to compare? :D

**Review by alfredsgenkins — CHANGES_REQUESTED** [review]

File: src/app/component/Slider/Slider.style.scss (hunk 7 above)
Comment (zans-laksa) [reply src/app/component/Slider/Slider.style.scss:60]: I've merged a branch 3.2.5-rc in order to update product card design and put a compare button there.
This comes from the 0ddade4

File: src/app/component/InstallPrompt/InstallPrompt.component.js (hunk 8 above)
Comment (zans-laksa) [reply src/app/component/InstallPrompt/InstallPrompt.component.js:54]: Not related. Just couldn't close this prompt when was testing on the mobile device. Fixed it.

File: src/app/store/LinkedProducts/LinkedProducts.dispatcher.js (hunk 9 above)
Comment (zans-laksa) [reply src/app/store/LinkedProducts/LinkedProducts.dispatcher.js:77]: This is from the merged commit dcf541e

File: src/app/component/SliderWidget/SliderWidget.style.scss (hunk 10 above)
Comment (zans-laksa) [reply src/app/component/SliderWidget/SliderWidget.style.scss:13]: This is from the merged commit 6c98e8e

File: src/app/component/SliderWidget/SliderWidget.container.js (hunk 11 above)
Comment (zans-laksa) [reply src/app/component/SliderWidget/SliderWidget.container.js:78]: This is from the merged commit 87ac9c9

File: src/app/component/ProductCard/ProductCard.component.js (hunk 1 above)
Comment (zans-laksa) [reply src/app/component/ProductCard/ProductCard.component.js]: Fixed

File: src/app/component/ProductCompare/ProductCompare.component.js (hunk 2 above)
Comment (zans-laksa) [reply src/app/component/ProductCompare/ProductCompare.component.js]: Fixed

File: src/app/component/ProductCompare/ProductCompare.component.js (hunk 3 above)
Comment (zans-laksa) [reply src/app/component/ProductCompare/ProductCompare.component.js]: Fixed

File: src/app/component/ProductCompare/ProductCompare.container.js (hunk 4 above)
Comment (zans-laksa) [reply src/app/component/ProductCompare/ProductCompare.container.js]: Fixed

File: src/app/component/ProductCompare/ProductCompare.style.scss (hunk 5 above)
Comment (zans-laksa) [reply src/app/component/ProductCompare/ProductCompare.style.scss]: Fixed

File: src/app/component/ProductCompareButton/ProductCompareButton.style.scss (hunk 6 above)
Comment (zans-laksa) [reply src/app/component/ProductCompareButton/ProductCompareButton.style.scss]: Fixed

**Comment by alfredsgenkins** [comment]
Please add dependencies to:
- https://github.com/scandipwa/compare-graphql of version `^1.0.0`
- https://github.com/scandipwa/performance of version `^1.4.1

Then, we can merge this one!

**Comment by alfredsgenkins** [comment]
Check if at the time you are there, the version 1.0.0 is available of Packagist: https://packagist.org/packages/scandipwa/compare-graphql#dev-master if not, ping me!

**Comment by zans-laksa** [comment]
I've updated dependencies.
But packagist still doesn't show the tag.

**Review by lianastaskevica — APPROVED** [review]

**Review by alfredsgenkins — APPROVED** [review]
